yunwoo/programmers-0/소인수분해.js

```diff
@@ -0,0 +1,25 @@
+function solution(n) {
+  const result = [];
+
+  primeFactorization(n, result);
+
+  const set = new Set(result);
+
+  return [...set];
+}
+
+const primeFactorization = (num, arr) => {
+  while (num % 2 === 0) {
+    num = num / 2;
+    arr.push(2);
+  }
+
+  for (let i = 3; i <= Math.sqrt(num); i += 2) {
+    while (num % i === 0) {
+      arr.push(i);
+      num = num / i;
+    }
+  }
+
+  if (num > 2) arr.push(num);
+};
```

Comment:
```suggestion
function solution(n) {
  const primeSet = primeFactorization(n);

  return [...primeSet];
}

const primeFactorization = (num) => {
  const set = new Set();
    
  while (num % 2 === 0) {
    num /= 2;
    set.add(2);
  }

  for (let i = 3; i <= Math.sqrt(num); i += 2) {
    while (num % i === 0) {
      set.add(i);
      num /= i;
    }
  }

  if (num > 2) set.add(num);
  return set;
};
```
- 지금 primeFactorization 함수에 2번째 매개변수로 배열을 받고, 이를 push와 같은 메서드를 직접 바꾸고 있는데 이런게 바로 사이드 이펙트입니다. 이런건 지양해주시는게 좋습니다.
- 차라리 primeFactorization 내부에 set을 하나 생성하고 여기다 add해주고 마지막에 set을 리턴해줘서 solution에서는 리턴받은 set을 배열로 변환하면 좋을 것 같습니다 🙇‍♂️

Reply:
- 와 set을 내부에서 사용할 생각을 못했네요 너무 좋은 방법인거 같습니다!!
- 빈배열을 넣을 걸 예상하고 사용한 부분이긴 한데 혹시 이런 사이드 이펙트가 끼치는 안좋은 영향이 있을까요?? 🤔
- 다음에 함수를 구성할 때 인자를 직접적으로 변경하는 부분을 최대한 빼야할지 궁금합니다!!